Feat: 회원가입 & 로그인 구현

junga970 opened the pull request:
## 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
### **📌 AS-IS**

**1. 응답, 예외 처리**
- 기본적으로 ApiResponse으로 응답합니다.
- 예외 처리는 ResponseEntity<ErrorResponse>로 응답합니다.
  - exception마다 http 상태 코드가 달라서 고민하다가 ResponseEntity를 사용하기로 했습니다.
  - 가독성을 고려해서 ErrorResponse에 ResponseEntity를 리턴하는 함수를 만들었습니다.

<br>

**2. 회원가입 기능 추가**
- 이메일 중복 시 CustomException을 발생시킵니다.
- 비밀번호는 bcrypt를 이용해 암호화한 뒤 DB에 저장합니다.

<br>

**3. 로그인 기능 추가**
-  가입된 이메일인지 확인합니다.
-  입력된 비밀번호와 유저 비밀번호가 일치하는지 확인합니다.
-  위 조건을 다 통과하면 JWT 토큰을 발급해 전달합니다.

<br>


**4. 테스트 코드 추가 (통과된 테스트는 체크 표시)**

Service/UserServiceTest

- [x] 회원가입 실패 - 이메일 중복
- [x] 회원가입 성공
- [x] 로그인 실패 - 잘못된 비밀번호
- [x] 로그인 실패 - 이메일에 해당하는 사용자 없음
- [x] 로그인 성공

<br>


<img width="1747" alt="스크린샷 2023-01-02 오전 4 54 50" src="https://user-images.githubusercontent.com/111635735/210183152-48821984-ffa3-4e12-930d-6648dd0a8ee8.png">



```java
public DataResponse login(UserLogin request) {
        ...

        if(!passwordEncoder.matches(request.getPassword(), user.getPassword())){
            System.out.println(request.getPassword());
            System.out.println(user.getPassword());
            throw new CustomException(INVALID_PASSWORD, HttpStatus.UNAUTHORIZED);
        }

	...
    }
```
로그인 성공 테스트에서 왜 실패한 건지 이유를 모르겠습니다.
이미지와 같이 비밀번호 검사 부분에서 걸립니다. 
임의로 프린트를 찍어서 확인한 인코딩된 비밀번호를 db에 넣고 
실제 api 테스트를 하면 문제없이 로그인이 됩니다... 

그리고 통과한 테스트 코드도 저렇게 작성하는게 맞는지 잘 모르겠습니다..! 


### **📌 TO-BE**
- ~~로그인 성공 테스트 통과~~

### **📌 테스트**
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- [x] API 테스트

junga970 committed:
Feat: 회원가입 기능 추가

- #1

junga970 committed:
Feat: 로그인 기능 추가

- #1

junga970 committed:
Test: 회원가입 & 로그인 테스트 코드 추가

- #1

mingming-mentor commented on src/main/java/com/example/bookstore/dto/response/DataResponse.java:
```diff
@@ -0,0 +1,21 @@
+@Getter
+@Setter
+@SuperBuilder
+public class DataResponse extends BaseResponse{
```
DataResponse와 BaseResponse를 합쳐도 괜찮을 것 같습니다. (data is nullable 형태로요)
그리고 data의 경우 Object로 선언하시는 것보다 generics를 사용하시는 것을 추천드려요
그러면 응답형태에 대해서 data type=A로 리턴을 해야하는데 개발자의 실수로 B형태를 넣더라도 compile 오류로 잡아낼수가 있습니다. Object면 해당 캐치가 어렵게 됩니다

mingming-mentor commented on src/main/java/com/example/bookstore/dto/response/ValidityErrorResponse.java:
```diff
@@ -0,0 +1,23 @@
+@Getter
+@Setter
+@SuperBuilder
+public class ValidityErrorResponse extends ErrorResponse {
```
마찬가지로 ErrorResponse와 합쳐도 괜찮을 것 같아요

mingming-mentor commented on src/main/java/com/example/bookstore/service/UserService.java:
```diff
@@ -0,0 +1,69 @@
+    private final JwtTokenProvider jwtTokenProvider;
+    private final PasswordEncoder passwordEncoder;
+
+    public BaseResponse register(UserRegister request) {
```
service layer에서는 client가 사용하는 응답값으로 내려주는 것이 좋지 않습니다.
각 계층별로 사용하는 데이터 구조가 분리된 것이 layered architecture에 맞는 설계이기때문이고, client에서 사용하는 응답값 형태는 controller의 책임이기 때문이에요. 그래서 여기서는 register가 제대로 되었다면 void로 return하고 에러가 발생했다면 exception을 던져주는 형태로 반환되는 형태를 추천드립니다

mingming-mentor commented on src/main/java/com/example/bookstore/controller/UserController.java:
```diff
@@ -0,0 +1,35 @@
+    private final UserService userService;
+
+    @PostMapping("/register")
+    public ResponseEntity register(@RequestBody @Valid UserRegister request) {
```
반환값을 ResponseEntity가 아닌 작성하신 BaseResponse로 return해주어도 좋을 것 같습니다. 굳이 ResponseEntity로 변환해서 보여주지 않아도 될것 같아요. 둘이 담고 있는 정보가 동일해서요

junga970 replied:
@mingming-mentor 

제가 ResponseEntity를 사용한 이유는 http 상태 코드 때문이었는데요!
컨트롤러 부분은 @ResponseStatus 어노테이션으로 처리했습니다.
문제는 예외 처리 부분인데요.. exception마다 http 상태 코드가 달라서 어떻게 처리할지 고민하다가
가독성을 고려해서 ErrorResponse에 ResponseEntity를 리턴하는 메서드를 구현해 사용했습니다.
다른 방법이 있다면 알려주시면 감사하겠습니다!!

mingming-mentor replied:
ResponseEntity의 역할을 현재 BaseResponse로도 동일하게 처리가 가능하기때문에 BaseResponse로 반환하는 형태로 하셔도 무방하다는 뜻이었습니다! 현재 BaseResponse도 httpStatus를 가지고 있지 않나요? (없다면 추가하셔도 괜찮습니다. 혹은 BaseResponse를 좀 덜어내셔도 좋구요 요지는 ResponseEntity에 들어가는 데이터와 BaseResponse의 데이터가 중복되기때문에 하나로 합치면 좋겠다는 뜻이었습니다.) 

현재는 보니 깔끔하게 잘 수정되어있는 것 같네요! 수고하셨습니다. (추가적인 궁금증이 있으시다면 알려주세요)

junga970 replied:
넵 감사합니다!!

mingming-mentor commented on src/main/java/com/example/bookstore/service/UserService.java:
```diff
@@ -0,0 +1,69 @@
+        User user = userRepository.findByEmail(request.getEmail())
+                .orElseThrow(() -> new CustomException(INVALID_EMAIL, HttpStatus.UNAUTHORIZED));
+
+        if(!passwordEncoder.matches(request.getPassword(), user.getPassword())){
```
질문주신 내용과 관련이 있는 부분인데요, 회원가입시에는 password encoding시 Bcrypt.gensalt()로 랜덤한 salt를 받아 hashing해두고 로그인할때는 passwordEncoder.match로 비교를 하니 두개의 값이 당연히 맞지 않다고 나올 것 같아요.
회원가입시에도 `passwordEncoder.encode(request.getPassword())`로 encoding해서 넣어주시면 될 것 같습니다

junga970 replied:
@mingming-mentor 

제가 봐도 일관성이 없는 코드였던 거 같아 빠르게 수정했습니다..ㅎㅎ
하지만 여전히 "로그인 성공 테스트"만 통과가 안됩니다..! 
전과 똑같이 `if(!passwordEncoder.matches(request.getPassword(), user.getPassword()))` 
이 부분에서 걸립니다.
api테스트는 문제 없이 로그인 성공했습니다.

junga970 replied:
코드 수정하고  PasswordEncoder를 BCryptPasswordEncoder로 바꿔주니 통과했습니다.
감사합니다!!

mingming-mentor reviewed (approved):
안녕하세요 밍밍멘토입니다! 몇가지 리뷰 드렸어요. response의 경우 중복된 형태가 조금 있어보이는데 잘 구조화하여 코드를 조금 가독성 높게 간소화하면 어떨까 합니다! 질문주신내용도 리뷰 코멘트로 달아두었으니 확인해보시고 알려주세요! 그 이외에는 코드가 전체적으로 깔끔하네요 :) 수고 많으셨습니다

junga970 committed:
Refactor: response dto & UserService 코드 수정

1. BaseResponse, DataResponse 제거 -> ApiResponse로 통합
- 중복 제거

2. ValidityErrorResponse 제거 -> ErrorResponse 통합
- 중복 제거

3. UserService: register(), login() 메서드 리턴값 변경
- controller와 service 역할 분리

4. register(): 비밀번호 인코딩 코드 변경
- 로그인과 마찬가지로 passwordEncoder 사용

- #1

junga970 committed:
Test: UserService 수정

1. "로그인 성공" 테스트 수정
- 테스트 통과

- #1
